bindings/rust/s2n-tls/src/testing/s2n_tls.rs

```diff
@@ -970,4 +970,162 @@ mod tests {
+
+        let context = connection
+            .application_context::<TestApplicationContext>()
+            .unwrap();
+        assert_eq!(context.test_value, 1142);
+    }
+
+    /// Test that an application context can be modified.
+    #[test]
+    fn test_app_context_modify() {
+        struct TestApplicationContext {
+            test_value: u32,
+        }
+
+        let mut connection = connection::Connection::new_server();
+        connection.set_application_context(TestApplicationContext { test_value: 0 });
+
+        let context = connection
+            .application_context_mut::<TestApplicationContext>()
+            .unwrap();
+        context.test_value += 1;
+
+        let context = connection
+            .application_context::<TestApplicationContext>()
+            .unwrap();
+        assert_eq!(context.test_value, 1);
+    }
+
+    /// Test that an application context can be overridden.
+    #[test]
+    fn test_app_context_override() {
+        struct TestApplicationContext {
+            test_value: u32,
+        }
+
+        let mut connection = connection::Connection::new_server();
+        connection.set_application_context(TestApplicationContext { test_value: 1142 });
+
+        {
+            let context = connection
+                .application_context::<TestApplicationContext>()
+                .unwrap();
+            assert_eq!(context.test_value, 1142);
+        }
+
+        // Override the context with a new value.
+        connection.set_application_context(TestApplicationContext { test_value: 10 });
+
+        {
+            let context = connection
+                .application_context::<TestApplicationContext>()
+                .unwrap();
+            assert_eq!(context.test_value, 10);
+        }
+
+        // Test overriding to a different context type.
+        struct AnotherApplicationContext {
+            another_test_value: i16,
+        }
+        connection.set_application_context(AnotherApplicationContext {
+            another_test_value: -20,
+        });
+
+        let context = connection
+            .application_context::<AnotherApplicationContext>()
+            .unwrap();
+        assert_eq!(context.another_test_value, -20);
+    }
+
+    /// Test that a context of another type can't be retrieved.
+    #[test]
+    fn test_app_context_invalid_type() {
+        struct TestApplicationContext {}
+        struct AnotherApplicationContext {}
+
+        let mut connection = connection::Connection::new_server();
+        connection.set_application_context(TestApplicationContext {});
+
+        // A context type that wasn't set shouldn't be returned.
+        assert!(connection
+            .application_context::<AnotherApplicationContext>()
+            .is_none());
+
+        // Retrieving the correct type succeeds.
+        assert!(connection
+            .application_context::<TestApplicationContext>()
+            .is_some());
```

Comment:
I was trying to figure out how your implementation produced this behavior, and eventually figured out it was that downcast_mut and downcast_ref check the type for you, so you don't have to track the type manually. Maybe that's common enough knowledge that we don't need to document it, but it feels enough like witchcraft to me that I think it's at least worth a note and a link to the downcast documentation.